File: verify/stripe_test.go
package verify

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"net/http"
	"testing"
)

func computeStripeSignature(secret, timestamp, body string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(timestamp + "." + body))
	return hex.EncodeToString(mac.Sum(nil))
}

func TestStripeVerifyValidSignature(t *testing.T) {
	secret := "whsec_test"
	body := `{"type":"charge.succeeded"}`
	ts := "1616161616"
	sig := computeStripeSignature(secret, ts, body)

	v := NewStripeVerifier(secret)
	headers := http.Header{}
	headers.Set("Stripe-Signature", fmt.Sprintf("t=%s,v1=%s", ts, sig))

	ok, err := v.Verify([]byte(body), headers)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !ok {
		t.Fatal("expected valid signature to pass verification")
	}
}

func TestStripeVerifyInvalidSignature(t *testing.T) {
	secret := "whsec_test"
	body := `{"type":"charge.succeeded"}`
	ts := "1616161616"
	wrongSig := computeStripeSignature("wrong-secret", ts, body)

	v := NewStripeVerifier(secret)
	headers := http.Header{}
	headers.Set("Stripe-Signature", fmt.Sprintf("t=%s,v1=%s", ts, wrongSig))

	ok, err := v.Verify([]byte(body), headers)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if ok {
		t.Fatal("expected invalid signature to fail verification")
	}
}

func TestStripeVerifyMissingHeader(t *testing.T) {
	v := NewStripeVerifier("whsec_test")
	headers := http.Header{}

	_, err := v.Verify([]byte(`{}`), headers)
	if err == nil {
		t.Fatal("expected error for missing header")
	}
}

func TestStripeVerifyMalformedHeader(t *testing.T) {
	v := NewStripeVerifier("whsec_test")

	tests := []struct {
		name   string
		header string
	}{
		{"garbage", "not-a-valid-header"},
		{"no timestamp", "v1=abc123"},
		{"no v1 sig", "t=1616161616"},
		{"empty", ""},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			headers := http.Header{}
			headers.Set("Stripe-Signature", tc.header)

			_, err := v.Verify([]byte(`{}`), headers)
			if err == nil {
				t.Fatalf("expected error for malformed header %q", tc.header)
			}
		})
	}
}

func TestStripeVerifyMultipleV1Signatures(t *testing.T) {
	secret := "whsec_test"
	body := `{"type":"charge.succeeded"}`
	ts := "1616161616"
	goodSig := computeStripeSignature(secret, ts, body)
	badSig := "deadbeef"

	v := NewStripeVerifier(secret)
	headers := http.Header{}
	headers.Set("Stripe-Signature", fmt.Sprintf("t=%s,v1=%s,v1=%s", ts, badSig, goodSig))

	ok, err := v.Verify([]byte(body), headers)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !ok {
		t.Fatal("expected verification to pass when one of multiple v1 sigs matches")
	}
}

File: src/verify/stripe.test.ts
import { describe, it, expect } from "vitest";
import { createHmac } from "node:crypto";
import { StripeVerifier } from "./stripe.js";

/** Computes a valid Stripe webhook signature for testing. */
function computeStripeSignature(
  secret: string,
  timestamp: string,
  body: string
): string {
  return createHmac("sha256", secret)
    .update(`${timestamp}.${body}`)
    .digest("hex");
}

describe("StripeVerifier", () => {
  it("accepts a valid signature", () => {
    const secret = "whsec_test";
    const body = '{"type":"charge.succeeded"}';
    const ts = "1616161616";
    const sig = computeStripeSignature(secret, ts, body);

    const verifier = new StripeVerifier(secret);
    const headers = {
      "stripe-signature": `t=${ts},v1=${sig}`,
    };

    const result = verifier.verify(Buffer.from(body), headers);
    expect(result.error).toBeUndefined();
    expect(result.valid).toBe(true);
  });

  it("rejects an invalid signature", () => {
    const secret = "whsec_test";
    const body = '{"type":"charge.succeeded"}';
    const ts = "1616161616";
    const wrongSig = computeStripeSignature("wrong-secret", ts, body);

    const verifier = new StripeVerifier(secret);
    const headers = {
      "stripe-signature": `t=${ts},v1=${wrongSig}`,
    };

    const result = verifier.verify(Buffer.from(body), headers);
    expect(result.error).toBeUndefined();
    expect(result.valid).toBe(false);
  });

  it("returns error for missing header", () => {
    const verifier = new StripeVerifier("whsec_test");
    const result = verifier.verify(Buffer.from("{}"), {});

    expect(result.error).toBeDefined();
    expect(result.valid).toBe(false);
  });

  it("returns error for malformed headers", () => {
    const verifier = new StripeVerifier("whsec_test");

    const cases = [
      { name: "garbage", header: "not-a-valid-header" },
      { name: "no timestamp", header: "v1=abc123" },
      { name: "no v1 sig", header: "t=1616161616" },
    ];

    for (const tc of cases) {
      const headers = { "stripe-signature": tc.header };
      const result = verifier.verify(Buffer.from("{}"), headers);
      expect(result.error).toBeDefined();
    }
  });

  it("accepts when one of multiple v1 signatures matches", () => {
    const secret = "whsec_test";
    const body = '{"type":"charge.succeeded"}';
    const ts = "1616161616";
    const goodSig = computeStripeSignature(secret, ts, body);
    const badSig = "deadbeef";

    const verifier = new StripeVerifier(secret);
    const headers = {
      "stripe-signature": `t=${ts},v1=${badSig},v1=${goodSig}`,
    };

    const result = verifier.verify(Buffer.from(body), headers);
    expect(result.error).toBeUndefined();
    expect(result.valid).toBe(true);
  });
});
